File: Chapter02(Serial)/SerialCommChatPocketPC/SerialCommChatPocketPC/Form1.vb
Public Class Form1
    Private WithEvents serialPort As New IO.Ports.SerialPort

    Private Sub DataReceived( _
           ByVal sender As Object, _
           ByVal e As System.IO.Ports.SerialDataReceivedEventArgs) _
           Handles serialPort.DataReceived

        txtReceivedMessage.BeginInvoke(New _
           myDelegate(AddressOf updateTextBox), _
           New Object() {})
    End Sub

    Public Delegate Sub myDelegate()
    Public Sub updateTextBox()
        txtReceivedMessage.Text = _
           serialPort.ReadExisting & _
           txtReceivedMessage.Text
    End Sub

    Private Sub MenuItem1_Click( _
       ByVal sender As System.Object, _
       ByVal e As System.EventArgs) _
       Handles MenuItem1.Click
        Try
            serialPort.WriteLine(txtMessageToSend.Text)
            txtReceivedMessage.Text = ">" & _
               txtMessageToSend.Text & vbCrLf & _
               txtReceivedMessage.Text
            txtMessageToSend.Text = String.Empty
        Catch ex As Exception
            MsgBox(ex.ToString)
        End Try
    End Sub

    Private Sub btnConnect_Click( _
       ByVal sender As System.Object, _
       ByVal e As System.EventArgs) _
       Handles btnConnect.Click
        Try
            If serialPort.IsOpen Then
                serialPort.Close()
            End If
            With serialPort
                .PortName = cbbCOMPorts.Text
                .BaudRate = 9600
                .Parity = IO.Ports.Parity.None
                .DataBits = 8
                .StopBits = IO.Ports.StopBits.One
            End With
            serialPort.Open()
            MsgBox("Port opened successfully!")
        Catch ex As Exception
            MsgBox(ex.ToString)
        End Try
    End Sub
End Class

File: Chapter02(Serial)/SerialCommChatPocketPC_CS/SerialCommChatPocketPC_CS/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

namespace SerialCommChatPocketPC_CS
{
    public partial class Form1 : Form
    {
        private System.IO.Ports.SerialPort serialPort =
            new System.IO.Ports.SerialPort();

        public Form1()
        {
            InitializeComponent();
        }

        private void DataReceived(object sender, System.IO.Ports.SerialDataReceivedEventArgs e)
        {
            txtReceivedMessage.BeginInvoke(new myDelegate(updateTextBox));
        }

        public delegate void myDelegate();
        public void updateTextBox()
        {
            //---for receiving plan ASCII text---
            txtReceivedMessage.Text = (serialPort.ReadExisting()) + txtReceivedMessage.Text;
            txtReceivedMessage.ScrollToCaret();
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            serialPort.DataReceived +=
               new System.IO.Ports.SerialDataReceivedEventHandler(DataReceived);
        }

        private void MenuItem1_Click(object sender, EventArgs e)
        {
            try
            {
                serialPort.Write(txtMessageToSend.Text + "\r");
                txtReceivedMessage.Text = ">" + txtMessageToSend.Text + "\r" + txtReceivedMessage.Text;
                txtMessageToSend.Text = string.Empty;
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.ToString());
            }
        }

        private void btnConnect_Click(object sender, EventArgs e)
        { 
            if (serialPort.IsOpen)
            {
                serialPort.Close();
            }
            try
            {
                serialPort.PortName = cbbCOMPorts.Text;
                serialPort.BaudRate = 9600;
                serialPort.Parity = System.IO.Ports.Parity.None;
                serialPort.DataBits = 8;
                serialPort.StopBits = System.IO.Ports.StopBits.One;
                serialPort.Open();
                MessageBox.Show("Port opened successfully!");
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.ToString());
            }
        }
    }
}